Major extension

Kucharssim opened the pull request:
Adds new distributions, as a by-product:
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1143
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1094

Does not add any functionality (apart from fixing some issues with the code here and there) - adds only new distributions. Adding new features will come after this update.

Distributions added:
* [x] Stretched beta
* [x] Generalized normal
* [x] Skew normal
* [x] Skew t
* [x] Generalized skewed t
* [x] Amoroso
* [x] beta-binomial
* [x] mixture of normal and uniform
* [x] mixture of normal and normal
* [x] zero-inflated poisson
* [x] zero-inflated negative binomial
* [x] Pareto
* [x] Gumbel
* [x] Gompertz
* [x] Beta prime
* [x] Wald
* [x] Frechet
* [x] Uniform
* [x] Geometric
* [x] Hypergeometric
* [x] Cauchy
* [x] Laplace
* [x] Loglogistic
* [x] Triangular
* [x] Weibull
* [x] Noncentral t
* [x] Scaled-shifted Student's t

koenderks reviewed (changes requested):
Let me start by saying that I love how each interface in your module is standardized with the same options and output!

This is what I did to review this PR. Since it is so much code and you said 99% is already existing code I mainly checked the functionality of each new distribution in JASP itself. To summarize, for all 27 new distributions I: 
- adjusted the parameters,
- checked all checkboxes for responsiveness and output,
- generated and fitted a new column of data and checked all available options there.

So therefore I guess this is more of a user-experience review instead of a code review. Took a long time but here are the main things I discovered (in no specific order):

- [ ] There are no help files for the new distributions yet.
- [ ] Is there any specific order to the distributions in the module menu? I get that the normal distribution is on top, but it might be useful to put them in alphabetical order or something so that the user can easily find the one they are looking for. Ultimately your call of course.
- [x] One thing I noticed when testing the Stretched Beta is that when I retype the upper bound on the `Range of x from ... to ...` as 1 (it starts as 1 as well), the option is highlighted in red, even though it is a valid value. I believe this happens in most of the distributions where you can specify this range.
![image](https://user-images.githubusercontent.com/25059399/107515020-b928c000-6baa-11eb-80e8-4802c83d843f.png)
While on that topic, some options stay red after being filled in, even though they are well within the correct range. Like this value of n = 0.2 in the Zero-inflated poison:
![image](https://user-images.githubusercontent.com/25059399/107521242-97cbd200-6bb2-11eb-921d-43df37e67f81.png)
And also this value of 8 in the log-logistic distribution:
![image](https://user-images.githubusercontent.com/25059399/107526666-60602400-6bb8-11eb-9707-0feac78f2596.png)
- [ ] I also noticed an encoded column name that slipped into the y-axis of the empirical cumulative distribution of the stretched beta. This is actually the case in all these plots, not just the new ones. You should probably call `decodeColNames()` on the name of the variable in `ldFillPlotECDF()`. 
![image](https://user-images.githubusercontent.com/25059399/107515280-1886d000-6bab-11eb-84b6-209d152d2576.png)
- [ ] I noted that the y-axis is a little bit off in the Q-Q plot of the stretched beta (first picture). In other q-q- plots the axes also seem to be a bit off, for example, an x-axis not going completely to the right (second picture).
![image](https://user-images.githubusercontent.com/25059399/107515530-6e5b7800-6bab-11eb-9f7a-718136fd4fcf.png)
![image](https://user-images.githubusercontent.com/25059399/107516659-060d9600-6bad-11eb-98aa-4e8662c52ecb.png)
- [ ] The interval options in the Generalized normal (version 1) look like they are disabled when any of the above options are checked, but they are not (as you know). I was a little bit confused because I thought that I could not click them initially. This is also the case in the other distributions. It is probably intended, but it would maybe be clearer that the user can interact with these options if they were not grayed out.
![image](https://user-images.githubusercontent.com/25059399/107515758-bbd7e500-6bab-11eb-9bb3-3a29fc63f738.png)
- [ ] I get this error in the Skewed generalized t distribution:
![image](https://user-images.githubusercontent.com/25059399/107518537-6a315980-6baf-11eb-927c-2736cb55db83.png)
Here is the .jasp file. I have weird options though.
[skewed_generalized_t_error.zip](https://github.com/jasp-stats/jaspDistributions/files/5958579/skewed_generalized_t_error.zip)
- [ ] It is a little bit counterintuitive that you have to click outside of the "Variable name" box before it generates the variable. Because if I just fill in a variable name and then click "Draw samples", nothing happens. This is more a general JASP thing though because we have more analyses that suffer from this issue (I know at least of some audit stuff that does).
- [ ] The probability mass plot of the beta-binomial distributions in this .jasp file looks a bit off. Shouldn't it range from 0 - 1000? I think it is because the options for the range are automatically adjusted when I adjust the number of trials, but not actually passed onto R so it does not change the figure axes.
![image](https://user-images.githubusercontent.com/25059399/107519481-9b5e5980-6bb0-11eb-8352-2e7103c91e22.png)
- [ ] The x-axis in the Histogram vs. theoretical PMF becomes very cluttered when the figure is so small. May I suggest a little less x-breaks in this plot specifically?
![image](https://user-images.githubusercontent.com/25059399/107520077-4ec74e00-6bb1-11eb-9ecc-c5a938c1a9cb.png)
- [ ] I know these things are hard to place, but the text is a little bit off in this plot of the mixture of normal and normal:
![image](https://user-images.githubusercontent.com/25059399/107520516-cb5a2c80-6bb1-11eb-85d8-f0736a5cd95a.png)
[normalnormal.zip](https://github.com/jasp-stats/jaspDistributions/files/5958639/normalnormal.zip)
- [ ] I notice that I am not able to obtain any confidence intervals and std.errors for the Pareto distribution. It gives a note under the table, but maybe it would be better to disable these options at all for this distribution.
![image](https://user-images.githubusercontent.com/25059399/107521984-7ae3ce80-6bb3-11eb-8cf3-ca9b70d0fcb6.png)
[pareto.zip](https://github.com/jasp-stats/jaspDistributions/files/5958730/pareto.zip)
- [ ] Isn't the option "Estimates" under "Estimate parameters" in every distribution conditional on the option "maximum likelihood"? If so, a suggestion would be to gray it out if maximum likelihood is not checked.

Very nice additions furthermore! You can see what you do with the feedback, but it is probably good to decode the axis name in the cdf. If you want me to look further at the code I can do that as well, but figured that would be okay for now since it was 99% the same.

Kucharssim commented:
Thanks @koenderks, the review was really helpful!!

I fixed some of the issues you mentioned, but not all - I comment on them below.

> There are no help files for the new distributions yet.

I'll do that in a separate PR, I created an issue for this so that everyone can shame me as long as I don't get it done. https://github.com/jasp-stats/INTERNAL-jasp/issues/1253

> Is there any specific order to the distributions in the module menu? I get that the normal distribution is on top, but it might be useful to put them in alphabetical order or something so that the user can easily find the one they are looking for. Ultimately your call of course.

They are loosely organised by their similarity, but I agree that it's not clear cut and obvious how that should determine the order. On the other hand, I don't really like the alphabetical order either, as it puts clearly related distributions very far away from each other (e.g. beta and stretched beta, gamma and inverse gamma, exponential and laplace, etc). Perhaps I could make additional groups: say for "continuous unbounded", "continuous semi-bounded", "continous-bounded"? Not sure whether that would be clear either, but at least it would add a little bit more structure. Maybe @EJWagenmakers has some ideas? The current order is here:
![Screenshot 2021-02-11 at 19 57 06](https://user-images.githubusercontent.com/7093546/107684687-5cf59700-6ca3-11eb-9bf6-7b087b7754f9.png)

> I get this error in the Skewed generalized t distribution:

I cannot reproduce this 😦 

> It is a little bit counterintuitive that you have to click outside of the "Variable name" box before it generates the variable. Because if I just fill in a variable name and then click "Draw samples", nothing happens. This is more a general JASP thing though because we have more analyses that suffer from this issue (I know at least of some audit stuff that does).

Yeah I agree, it's little bit weird behaviour. Maybe something for @boutinb ? I'll make an issue for you...

> The probability mass plot of the beta-binomial distributions in this .jasp file looks a bit off. Shouldn't it range from 0 - 1000? I think it is because the options for the range are automatically adjusted when I adjust the number of trials, but not actually passed onto R so it does not change the figure axes.

For now, I fixed it by not automatically adjusting the range depending on the parameter option. I'll again make an issue for @boutinb to look into...
